packages/client/src/pages/page-editor/page-editor.script-block.vue

```diff
@@ -56,177 +56,139 @@
+async function changeType() {
+	const { canceled, result: type } = await os.select({
+		title: i18n.ts._pages.selectType,
+		groupedItems: getScriptBlockList!(props.getExpectedType ? getExpectedType() : null)
+	});
+	if (canceled) return;
 
-		this.$watch(() => this.modelValue.type, (t) => {
-			this.warn = null;
+	props.modelValue.type = type;
+}
 
-			if (this.modelValue.type === 'fn') {
-				const id = uuid();
-				this.modelValue.value = {
-					slots: [],
-					expression: { id, type: null }
-				};
-				return;
-			}
+function _getExpectedType(slot: number) {
+	return props.hpml.getExpectedType(props.modelValue, slot);
+}
 
-			if (this.modelValue.type && this.modelValue.type.startsWith('fn:')) {
-				const fnName = this.modelValue.type.split(':')[1];
-				const fn = this.hpml.getVarByName(fnName);
+if (props.modelValue.value == null) props.modelValue.value = null;
 
-				const empties = [];
-				for (let i = 0; i < fn.value.slots.length; i++) {
-					const id = uuid();
-					empties.push({ id, type: null });
-				}
-				this.modelValue.args = empties;
-				return;
-			}
+if (props.modelValue.value && props.modelValue.value.slots) slots = props.modelValue.value.slots.map(x => x.name).join('\n');
 
-			if (isLiteralValue(this.modelValue)) return;
+watch(() => props.modelValue.type, (t) => {
+	warn = null;
 
-			const empties = [];
-			for (let i = 0; i < funcDefs[this.modelValue.type].in.length; i++) {
-				const id = uuid();
-				empties.push({ id, type: null });
-			}
-			this.modelValue.args = empties;
+	if (props.modelValue.type === 'fn') {
+		const id = uuid();
+		props.modelValue.value = {
+			slots: [],
+			expression: { id, type: null }
+		};
+		return;
+	}
 
-			for (let i = 0; i < funcDefs[this.modelValue.type].in.length; i++) {
-				const inType = funcDefs[this.modelValue.type].in[i];
-				if (typeof inType !== 'number') {
-					if (inType === 'number') this.modelValue.args[i].type = 'number';
-					if (inType === 'string') this.modelValue.args[i].type = 'text';
-				}
-			}
-		});
+	if (props.modelValue.type && props.modelValue.type.startsWith('fn:')) {
+		const fnName = props.modelValue.type.split(':')[1];
+		const fn = props.hpml.getVarByName(fnName);
 
-		this.$watch(() => this.modelValue.args, (args) => {
-			if (args == null) {
-				this.warn = null;
-				return;
-			}
-			const emptySlotIndex = args.findIndex(x => x.type === null);
-			if (emptySlotIndex !== -1 && emptySlotIndex < args.length) {
-				this.warn = {
-					slot: emptySlotIndex
-				};
-			} else {
-				this.warn = null;
-			}
-		}, {
-			deep: true
-		});
+		const empties: any[] = [];
+		for (let i = 0; i < fn.value.slots.length; i++) {
+			const id = uuid();
+			empties.push({ id, type: null });
+		}
+		props.modelValue.args = empties;
+		return;
+	}
 
-		this.$watch(() => this.hpml.variables, () => {
-			if (this.type != null && this.modelValue) {
-				this.error = this.hpml.typeCheck(this.modelValue);
-			}
-		}, {
-			deep: true
-		});
```

Comment:
So while adding it back in I remembered.

```ts
if (this.type != null && this.modelValue) {
  this.error = this.hpml.typeCheck(this.modelValue);
}
```

`this.type` is always null, because it doesn't exist in the scope of this component. This causes the watch to do nothing, rendering it useless.

Reply:
I'm not sure of the original intentions, but it is probably a mistake and meant to refer to something else. Probably either `modelValue.type` or `getExpectedType()`?

However, looking back at the history, the code was already introduced with `this.type != null` and no definition of `type` in 05b8111c1906c1285c9ddde758eda45b83792244 by @syuilo. (To be more precise in <https://github.com/misskey-dev/misskey/pull/4811/commits/ac3bf1ec254de94548ee6f845756de499d84329c#diff-c00afbbc9270ec376c6a5d05b486cfe022527d1336727bb678029e133a51b53d>)